[NWS] added Current, Daily; improved codes and texts

chunshek opened the pull request:
I have improved NWS source as follows:
- Implemented Current for Main and Secondary
- Improved alert text by splitting alert description and instruction, and removing unsightly line breaks within sentences.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsResultConverter.kt:
```diff
@@ -182,6 +215,7 @@ private fun getDailyForecast(
 
 fun getAlerts(alerts: List<NwsAlert>?): List<Alert>? {
     if (alerts.isNullOrEmpty()) return null
+    val regex = Regex("""([0-9A-Za-z.,]) ?\n([0-9A-Za-z])""")
```
So, this does replace only the linebreaks that were mistakenly added? And you differentiate them because they are preceeded by a space, did I understand right?

papjul replied:
Why not a simple `.replace(" \n", " ")`?

chunshek replied:
What I was really checking for was a **single** `\n`, surrounded by "word characters" (i.e. `[0-9A-Za-z.,]`) on both sides.  In other words, multiple line breaks would not be replaced.

The space was optional, hence the `?`. But if I replace `" ?\n"` with `" "` then all the line breaks would disappear.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -99,8 +101,10 @@ class NwsService @Inject constructor(
+            .getOrElse(id) { null }?.getOrElse("station") { null }
 
-        if (gridId.isNullOrEmpty() || gridX.isNullOrEmpty() || gridY.isNullOrEmpty()) {
+        if (gridId.isNullOrEmpty() || gridX.isNullOrEmpty() || gridY.isNullOrEmpty() || station.isNullOrEmpty()) {
```
It is only an issue if requesting `CURRENT`

chunshek replied:
Will add extra condition.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -154,16 +173,60 @@ class NwsService @Inject constructor(
+        val station = location.parameters
+            .getOrElse(id) { null }?.getOrElse("station") { null }
+        if (station.isNullOrEmpty()) {
+            return Observable.error(InvalidLocationException())
```
Same here

chunshek replied:
Will add extra condition.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -154,16 +173,60 @@ class NwsService @Inject constructor(
-            convertSecondary(it)
+        val failedFeatures = mutableListOf<SourceFeature>()
+        val current = if (requestedFeatures.contains(SourceFeature.FEATURE_CURRENT)) {
+            mApi.getCurrent(
```
Do that if `!stationId.isNullOrEmpty()` then in the else add to `failedFeatures`

chunshek replied:
I have added the if-then-else.

But if we had caught this condition with `InvalidLocationException()` already at the start of the function, in theory we would never reach this failure here.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -202,10 +265,13 @@ class NwsService @Inject constructor(
             .getOrElse(id) { null }?.getOrElse("gridX") { null }
         val currentGridY = location.parameters
             .getOrElse(id) { null }?.getOrElse("gridY") { null }
+        val currentStation = location.parameters
```
Only with `CURRENT`

chunshek replied:
Added extra condition.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -202,10 +265,13 @@ class NwsService @Inject constructor(
             currentGridX.isNullOrEmpty() ||
-            currentGridY.isNullOrEmpty()
+            currentGridY.isNullOrEmpty() ||
+            currentStation.isNullOrEmpty()
```
Same

chunshek replied:
Added extra condition.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -220,10 +286,20 @@ class NwsService @Inject constructor(
             if (it.properties == null) {
                 throw InvalidLocationException()
             }
+            val stations = mApi.getStations(
```
Same

chunshek replied:
Added extra condition.

chunshek replied:
`requestLocationParameters` doesn't know which features are being requested.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/nws/NwsService.kt:
```diff
@@ -220,10 +286,20 @@ class NwsService @Inject constructor(
                 "gridX" to it.properties.gridX.toString(),
-                "gridY" to it.properties.gridY.toString()
+                "gridY" to it.properties.gridY.toString(),
+                "station" to stations.features.first().properties?.stationIdentifier!!
```
Only there if requesting `CURRENT`

chunshek replied:
`requestLocationParameters` doesn't know which features are being requested.

chunshek commented:
Amended the PR. Here's a summary of the changes:
- Implemented Current for Main and Secondary
- Added official daily forecasts from NWS rather than generating from hourly list
- Added interpretation of `weatherCode` for current and daily forecasts
- Added interpretation of `weatherText` and `weatherCode` for hourly forecasts based on NWS usage
- Improved alert text by splitting description and instruction, and removing unsightly line breaks within sentences

chunshek commented:
For `weatherText`, we can use the message in the API output directly for current and daily, but have to construct them for hourly.

Question: I am currently using a mix of resource strings for weather conditions and hardcoded strings. Do you prefer all the hardcoded strings to be added to the resource strings? If so, would you like them marked with `translatable="false"` or let translators have at it?

papjul commented:
Yes, please create non-translatable resource strings. We never know if/when they become shared by multiple sources.

chunshek commented:
OK. I'm on it.

papjul commented:
Thanks! I made a few changes in 5ae5077acaee7ab97b5595ac393bf8883d4bcc3a
